lib/resourcebuilder/apps.go

```diff
@@ -86,41 +91,77 @@ func waitForDeploymentCompletion(ctx context.Context, client appsclientv1.Deploy
+		var progressingCondition *appsv1.DeploymentCondition
+		var replicafailureCondition *appsv1.DeploymentCondition
+		for idx, dc := range d.Status.Conditions {
 			switch dc.Type {
-			case appsv1.DeploymentProgressing, appsv1.DeploymentAvailable:
-				if dc.Status == "False" {
-					deploymentConditions = append(deploymentConditions, fmt.Sprintf(", reason: %s, message: %s", dc.Reason, dc.Message))
-				}
+			case appsv1.DeploymentProgressing:
+				progressingCondition = &d.Status.Conditions[idx]
+			case appsv1.DeploymentAvailable:
+				availableCondition = &d.Status.Conditions[idx]
 			case appsv1.DeploymentReplicaFailure:
-				if dc.Status == "True" {
-					deploymentConditions = append(deploymentConditions, fmt.Sprintf(", reason: %s, message: %s", dc.Reason, dc.Message))
-				}
+				replicafailureCondition = &d.Status.Conditions[idx]
+			}
+		}
+
+		if replicafailureCondition != nil && replicafailureCondition.Status == corev1.ConditionTrue {
+			lastErr = &payload.UpdateError{
+				Nested:  fmt.Errorf("deployment %s has some pods failing; unavailable replicas=%d", iden, d.Status.UnavailableReplicas),
+				Reason:  "WorkloadNotProgressing",
+				Message: fmt.Sprintf("deployment %s has a replica failure %s: %s", iden, replicafailureCondition.Reason, replicafailureCondition.Message),
+				Name:    iden,
+			}
+			return false, nil
+		}
+
+		if availableCondition == nil && availableCondition.Status == corev1.ConditionFalse {
+			lastErr = &payload.UpdateError{
+				Nested:  fmt.Errorf("deployment %s is not available; updated replicas=%d of %d, available replicas=%d of %d", iden, d.Status.UpdatedReplicas, d.Status.Replicas, d.Status.AvailableReplicas, d.Status.Replicas),
+				Reason:  "WorkloadNotAvailable",
+				Message: fmt.Sprintf("deployment %s is not available %s: %s", iden, availableCondition.Reason, availableCondition.Message),
+				Name:    iden,
+			}
+			return false, nil
+		}
+
+		if progressingCondition == nil && progressingCondition.Status == corev1.ConditionTrue {
+			lastErr = &payload.UpdateError{
+				Nested:  fmt.Errorf("deployment %s is progressing; updated replicas=%d of %d, available replicas=%d of %d", iden, d.Status.UpdatedReplicas, d.Status.Replicas, d.Status.AvailableReplicas, d.Status.Replicas),
+				Reason:  "WorkloadNotAvailable",
+				Message: fmt.Sprintf("deployment %s is progressing %s: %s", iden, progressingCondition.Reason, progressingCondition.Message),
+				Name:    iden,
 			}
+			return false, nil
 		}
 
-		klog.V(4).Infof("Deployment %s is not ready. status: (replicas: %d, updated: %d, ready: %d, unavailable: %d%s)",
-			d.Name,
-			d.Status.Replicas,
-			d.Status.UpdatedReplicas,
-			d.Status.ReadyReplicas,
-			d.Status.UnavailableReplicas,
-			strings.Join(deploymentConditions, ""))
+		klog.Errorf("deployment %s is in unknown state", iden)
 		return false, nil
 	}, ctx.Done())
+	if err != nil {
+		if err == wait.ErrWaitTimeout && lastErr != nil {
+			return lastErr
```

Comment:
would it be worth dumping the deployment yaml, especially if we're in the "deployment is in unknown state" case?

It's a tradeoff between messy logs and making people jump through more hoops to dig up debug info.

Reply:
I don't think printing the entire YAML is the right thing to do, we shouldn't rely on this to provide the entire debugging, we have the `oc adm must-gather` to give us more details.

Reply:
I don't think we should dump the YAML unless in all situations we're confident that would immediately make root cause clear. Is there an example of other higher level tools doing such in the product already?

Reply:
I can't think of one offhand, but is there any scenario in which an install fails with this error message and the user does not immediately have to go look up that yaml to figure out the next step?

Reply:
I think in all situations we are better served if we use bootstrap gather and must-gather to collect failure logs.

Also looking at the entire object wouldn't necessarily give user what's failing. If there is a specific section that provides more values that _might_ be more appropriate.

Reply:
I think bootstrap gather and must-gather are really only useful to support and engineering. We don't provide any guidance on what to do with that data which is something that I think we need to consider as an organization going forward.

In my opinion this PR as is does a decent job making it clear where next to look. I worry that dumping YAML is both noisy and shifts more debugging into the CVO and Installer. Lets go with things as is and see how that helps.